awxkit/awxkit/cli/custom.py

```diff
@@ -196,6 +197,116 @@ def perform(self, slug):
         return self.page.get()
 
 
+class RoleMixin(object):
+
+    has_roles = [
+        ['organizations', 'organization'],
+        ['projects', 'project'],
+        ['inventories', 'inventory'],
+        ['inventory_scripts', 'inventory_script'],
+        ['teams', 'team'],
+        ['credentials', 'credential'],
+        ['job_templates', 'job_template'],
+        ['workflow_job_templates', 'workflow_job_template'],
+    ]
+    roles = {}  # this is calculated once
+
+    def add_arguments(self, parser):
+
+        if not RoleMixin.roles:
+            for resource, flag in self.has_roles:
+                options = self.page.__class__(
+                    self.page.endpoint.replace(self.resource, resource),
+                    self.page.connection
+                ).options()
+                RoleMixin.roles[flag] = [
+                    role.replace('_role', '')
+                    for role in options.json.get('object_roles', [])
+                ]
+
+        possible_roles = set()
+        for v in RoleMixin.roles.values():
+            possible_roles.update(v)
+
+        resource_group = parser.choices[self.action].add_mutually_exclusive_group(
+            required=True
+        )
+        parser.choices[self.action].add_argument(
+            'id', type=int,
+            help='The ID of the {} to {} access to/from'.format(
+                self.resource, self.action
+            )
+        )
+        for _type in RoleMixin.roles.keys():
+            if _type == 'team' and self.resource == 'team':
+                # don't add a team to a team
+                continue
+            resource_group.add_argument(
+                '--{}'.format(_type), type=int, metavar='ID',
+                help='The ID of the target {}'.format(_type),
+            )
+        parser.choices[self.action].add_argument(
+            '--type', type=str, choices=possible_roles, required=True,
+            help='The name of the role to {}'.format(self.action)
+        )
+
+    def perform(self, **kwargs):
+        for resource, flag in self.has_roles:
```

Comment:
If I'm reading this loop right, then if I supply `--inventory 4 --project 5 --type admin`, then it will go on its merry way granting admin permission to inventory 4 as well as project 5... baring any errors.

I don't think it should do that. Validation could be added here (I'd rather not). That kind of brings me back to why I think this would be better to take everything as fully positional, like `awx user grant 15 member organization 100`.

Also, because being lightweight and developer-friendly is such a big deal for this CLI, I'd actually prefer a way to use it with the role id. Say, `awx user grant 15 4032`, so if they user wants to go look up organization 100's member role primary key... go right ahead.

Reply:
`--inventory 4 --project 5` these resource flags are mutually exclusive:

```
~ awx users grant 2 --inventory 4 --project 5 --type admin
awx users grant: argument --project: not allowed with argument --inventory
```

Reply:
https://github.com/ansible/awx/pull/4630/files#diff-a8a950fa8ea340c107209be099b3710aR231

Reply:
As wordy this is:

`~ awx user grant 15 --organization N --type admin`

I think:

`~ awx user grant 15 4032`

...is going to be incredibly cumbersome for users to figure how to get `4032` and understand that it means e.g., "admin for <Org> N"

Reply:
What if the type were positional, i.e.,

`~ awx grant 15 admin --organization N`

Reply:
well, the user is also going to want to do `awx user grant fred`, not `awx user grant 15`. But yes, forcing users to look up the role IDs is not practical (and not portable).

Reply:
I would still expect there to be a variation of the role association command starting from the `awx role` root command, because from that perspective it behaves as many-to-many like all the others. While that might not exist now, I hope we agree on the principle and that it could be added. Ultimately a `grant` command is a shortcut.

Reply:
@AlanCoding I could see an argument at some point for something like:

`awx role grant NNNN --user 1 `
`awx role revoke NNNN --user 1 `